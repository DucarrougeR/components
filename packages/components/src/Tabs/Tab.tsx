import { ReactNode } from 'react'
import { rgba } from 'polished'
import styled, { css } from 'styled-components'
import {
  reset,
  SpaceProps,
  LayoutProps,
  BorderProps,
  space,
  TypographyProps,
  typography,
  layout,
  border,
} from 'looker-design-tokens'

export interface TabProps
  extends BorderProps,
    LayoutProps,
    SpaceProps,
    TypographyProps {
  children: ReactNode
  selected?: boolean
  disabled?: boolean
  onSelect?: () => void
}

const tabFocusRing = (color: string) => css`
  box-shadow: 0 0 0 0.15rem ${rgba(color, 0.25)};
  outline: none;
`

export const Tab = styled.button.attrs((props: TabProps) => ({
  minWidth: '3rem',
  onClick: () => {
    if (!props.disabled && props.onSelect) {
      props.onSelect()
    }
  },
}))<TabProps>`
  ${reset}
  ${space}
  ${layout}
  ${border}
  ${typography}

  background: transparent;
  border-bottom: 3px solid;
  border-bottom-color: ${props =>
    props.selected ? props.theme.colors.palette.purple400 : 'transparent'};
  border-radius: 0;
  color: ${props =>
    props.selected
      ? props.theme.colors.palette.charcoal800
      : props.theme.colors.palette.charcoal500};
  cursor: pointer;

  & + & {
    margin-left: ${props => props.theme.space.xlarge};
  }

  &:hover {
    border-bottom-color: ${props =>
      props.selected
        ? props.theme.colors.palette.purple400
        : props.theme.colors.palette.charcoal300};
  }

  &:active {
    border-bottom-color: ${props =>
      props.selected
        ? props.theme.colors.palette.purple400
        : props.theme.colors.palette.charcoal400};
  }

  &:focus {
    ${props => tabFocusRing(props.theme.colors.palette.purple300)}
  }

  &:disabled {
    color: ${props => props.theme.colors.palette.charcoal300};

    &:active,
    &:hover {
      border-bottom-color: transparent;
    }
  }
`

Tab.defaultProps = { fontWeight: 'semiBold', pb: 'small', pt: 'xsmall' }
