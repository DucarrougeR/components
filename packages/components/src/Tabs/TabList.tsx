import React, { cloneElement } from 'react'
import styled from 'styled-components'

export interface TabListProps {
  children: JSX.Element[]
  selectedIndex?: number
  onSelectTab?: (index: any) => void
}

export const TabList: React.FC<TabListProps> = ({
  children,
  selectedIndex,
  onSelectTab,
}) => {
  const clonedChildren = React.Children.map(
    children,
    (child: JSX.Element, index: number) => {
      return cloneElement(child, {
        onSelect: () => onSelectTab && onSelectTab(index),
        selected: index === selectedIndex,
        selectedIndex,
      })
    }
  )

  return <TabListContainer>{clonedChildren}</TabListContainer>
}

const TabListContainer = styled.div`
  border-bottom: 1px solid ${props => props.theme.colors.palette.charcoal200};
`
