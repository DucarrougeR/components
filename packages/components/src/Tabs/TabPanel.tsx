import React from 'react'

export interface TabPanelProps {
  selected?: boolean
}

export const TabPanel: React.FC<TabPanelProps> = ({ children, selected }) =>
  selected ? <>{children}</> : null
