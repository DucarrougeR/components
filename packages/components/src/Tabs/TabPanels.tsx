import React, { cloneElement } from 'react'
import styled from 'styled-components'
import { SpaceProps, space, reset } from 'looker-design-tokens'
import omit from 'lodash/omit'

export interface TabPanelsProps extends SpaceProps {
  children: JSX.Element[]
  selectedIndex?: number
  onSelectTab?: (index: number) => void
}

export const TabPanels: React.FC<TabPanelsProps> = ({
  children,
  selectedIndex,
  ...props
}) => {
  const spaceProps = omit(props, 'onSelectTab')

  const clonedChildren = React.Children.map(
    children,
    (child: JSX.Element, index: number) => {
      return cloneElement(child, {
        selected: index === selectedIndex,
      })
    }
  )

  return <Layout {...spaceProps}>{clonedChildren}</Layout>
}

const Layout = styled.div<SpaceProps>`
  ${reset}
  ${space}
`

Layout.defaultProps = { pt: 'large' }
