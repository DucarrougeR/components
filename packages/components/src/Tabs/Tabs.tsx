import React, { Children, cloneElement, FC, useRef, useState } from 'react'

export interface UseTabsProps {
  controlledIndex?: number
  defaultIndex?: number
  index?: number | undefined
  isControlled?: boolean
  onChange?: (index: number) => void
}

export interface TabsProps extends UseTabsProps {
  children: JSX.Element[]
}

export function useTabs(props?: UseTabsProps) {
  const defaultIndex = (props && props.defaultIndex) || 0
  const onChange = props && props.onChange
  const isControlled = (props && props.isControlled) || false

  const [selectedIndex, setSelectedIndex] = useState(defaultIndex)

  return {
    onSelectTab: (index: number) => {
      onChange && onChange(index)
      if (!isControlled) {
        setSelectedIndex(index)
      }
    },
    selectedIndex,
  }
}

export const Tabs: FC<TabsProps> = ({
  children,
  index: controlledIndex,
  defaultIndex,
  onChange,
}) => {
  const { current: isControlled } = useRef(controlledIndex !== undefined)

  /* eslint-disable no-console */
  if (!isControlled && controlledIndex !== undefined) {
    console.warn(
      'Tabs is changing from uncontrolled to controlled. Tabs should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in.'
    )
  }

  if (isControlled && controlledIndex === undefined) {
    console.warn(
      'Tabs is changing from uncontrolled to controlled. Tabs should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in.'
    )
  }
  /* eslint-enable no-console */

  const tabs = useTabs({ defaultIndex, isControlled, onChange })

  const clonedChildren = Children.map(children, (child: JSX.Element) => {
    return cloneElement(child, tabs)
  })

  return <>{clonedChildren}</>
}
