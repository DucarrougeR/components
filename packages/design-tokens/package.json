{
  "name": "looker-design-tokens",
  "license": "UNLICENSED",
  "version": "0.1.19-alpha.0",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "registry": "https://nexusrepo.looker.com/repository/looker-npm-private/"
  },
  "scripts": {
    "build:declaration": "tsc --project tsconfig.build.json",
    "build:es": "BABEL_ENV=build babel src --root-mode upward --out-dir dist --source-maps --extensions .ts,.tsx --delete-dir-on-start --no-comments"
  },
  "devDependencies": {
    "@types/lodash": "^4.14.136",
    "@types/styled-components": "^4.1.19",
    "@types/styled-system": "^5.1.2",
    "csstype": "^2.6.3",
    "lodash": "^4.17.15",
    "polished": "^3.4.1",
    "styled-components": "^4.3.2",
    "styled-system": "^5.1.2",
    "uuid": "^3.3.2"
  },
  "peerDependencies": {
    "@types/styled-components": "^4.1.19",
    "@types/styled-system": "^5.1.2",
    "styled-components": "^4.3.2",
    "styled-system": "^5.1.2"
  },
  "gitHead": "05deaab6fabe072f9754f7164520db0d0616cab0"
}
