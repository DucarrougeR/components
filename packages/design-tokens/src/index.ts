export * from './system'
export * from './theme'

export { palette, radii, semanticColors } from './tokens'

// Useful external helpers
export * from './customizeable_attributes'
export * from './helpers/animations'
