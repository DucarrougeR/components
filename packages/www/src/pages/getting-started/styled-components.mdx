---
name: Styled Components
---

import { VisitedLink } from './demos'

# Styled Components

[Styled Components](https://www.styled-components.com/) is a low level CSS
library used by Lens.

> ðŸ‘‹ Heads up! ðŸ‘‹
>
> **You should rarely need to use Styled Components.** Instead use the [`<Box/>`](/layout/box)
> component and its props which ensures you use Lens' theme values and
> stay within the Lens style guidelines. Styled Components should only be
> used for configuring [pseudo classes (eg `:hover`)](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-classes), [pseudo elements (eg
> `:before`)](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements), animations, or rarely when `<Box/>`'s properties are insufficient.

### Things You'll Learn

- Use Styled Components' `styled` function for styling pseudo classes and pseudo
  elements.
- Strip props out before passing them to lower level components to avoid runtime
  errors.
- Styled Components provides access to a component's prop values so styling can
  be computed dynamically.

## Styling a Lens Component

Let's pretend we want to build a new kind of `<Link>` that renders visited links
a bit differently. Our designer has laid out the following specs:

1.  Visited links should render red.
2.  The component should accept a prop called `visited` that renders the text as
    strike-through on hover.

To handle this, we'll use Styled Components' `styled` function because we'll
need to style both the `:visited` and `:hover` pseudo classes. Step one is to
set the `:visited` color to our palette.red400 color. We can access the Theme
via Styled Components, so this is easy enough:

<VisitedLink href="https://www.google.com">Google.com</VisitedLink>

```jsx static
import { Link } from '@looker/components'

const VisitedLink = styled(Link)`
  :visited {
    color: ${props => props.theme.colors.palette.red400};
  }
`
```

Note, `styled` expects to be passed a React component directly. In the example
above, we pass it the `Link` Lens component.

## Styling Child Components and Linter Errors

Styled Components [provides a useful mechanism for styling child components](https://www.styled-components.com/docs/advanced#referring-to-other-components). This syntax, unfortunately, [generates a known linter error](https://github.com/styled-components/stylelint-processor-styled-components/issues/34#issuecomment-325339118). To remove the linter error, simply add the `/* sc-selector */` tag to your style definition. For example:

```jsx static
import { styled, Link } from '@looker/components'

const Card = styled.div`
  ${/* sc-selector */ Link} {
    background-color: ${props => props.theme.palette.green300};
  }
`
```
