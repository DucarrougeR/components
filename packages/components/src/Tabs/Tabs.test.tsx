import 'jest-styled-components'
import '@testing-library/jest-dom/extend-expect'
import { shallow } from 'enzyme'
import {
  assertSnapshotShallow,
  mountWithTheme,
} from '@looker/components-test-utils'
import { theme } from '@looker/design-tokens'
import React from 'react'
import { ThemeProvider } from 'styled-components'
import { fireEvent, render } from '@testing-library/react'
import { Tab } from './Tab'
import { TabList } from './TabList'
import { TabPanel } from './TabPanel'
import { TabPanels } from './TabPanels'
import { Tabs, useTabs } from './Tabs'

test('Tabs snapshot works as expected', () => {
  assertSnapshotShallow(
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is tab3 content</TabPanel>
      </TabPanels>
    </Tabs>
  )
})

test('shows the correct number of navigation tabs', () => {
  const tabs = shallow(
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is tab3 content</TabPanel>
      </TabPanels>
    </Tabs>
  )
  expect(tabs.find('Tab')).toHaveLength(3)
})

test('starts with Tab at index 0 opened', () => {
  const tabs = mountWithTheme(
    <Tabs>
      <TabList>
        <Tab>tab1</Tab>
        <Tab>tab2</Tab>
        <Tab disabled>tab3</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>this is tab1 content</TabPanel>
        <TabPanel>this is tab2 content</TabPanel>
        <TabPanel>this is tab3 content</TabPanel>
      </TabPanels>
    </Tabs>
  )

  const EachTabPanel: any = tabs
    .find(TabPanels)
    .find(TabPanel)
    .map((panel: any) => panel.props())

  expect(EachTabPanel[0].children).toEqual('this is tab1 content')
  expect(EachTabPanel[0].selected).toEqual(true)
  expect(EachTabPanel[1].selected).toEqual(false)
  expect(EachTabPanel[2].selected).toEqual(false)
})

const ChangingPanel = () => {
  return (
    <ThemeProvider theme={theme}>
      <Tabs>
        <TabList>
          <Tab>tab1</Tab>
          <Tab>tab2</Tab>
          <Tab disabled>tab3</Tab>
        </TabList>
        <TabPanels>
          <TabPanel>this is tab1 content</TabPanel>
          <TabPanel>this is tab2 content</TabPanel>
          <TabPanel>this is tab3 content</TabPanel>
        </TabPanels>
      </Tabs>
    </ThemeProvider>
  )
}

test('clicking on tab opens correct panel', () => {
  const { getByText, queryByText } = render(<ChangingPanel />)

  expect(queryByText('this is tab1 content')).toBeInTheDocument()
  expect(queryByText('this is tab2 content')).not.toBeInTheDocument()
  fireEvent.click(getByText('tab2'))
  expect(queryByText('this is tab1 content')).not.toBeInTheDocument()
  expect(queryByText('this is tab2 content')).toBeInTheDocument()
})

const DisableTab = () => {
  return (
    <ThemeProvider theme={theme}>
      <Tabs>
        <TabList>
          <Tab>tab1</Tab>
          <Tab>tab2</Tab>
          <Tab disabled>tab3</Tab>
        </TabList>
        <TabPanels>
          <TabPanel>this is tab1 content</TabPanel>
          <TabPanel>this is tab2 content</TabPanel>
          <TabPanel>this is the disable tab-panel</TabPanel>
        </TabPanels>
      </Tabs>
    </ThemeProvider>
  )
}

test('clicking on disable tab does not change panel', () => {
  const { getByText, queryByText } = render(<DisableTab />)

  expect(queryByText('this is tab1 content')).toBeInTheDocument()
  expect(queryByText('this is the disable tab-panel')).not.toBeInTheDocument()
  fireEvent.click(getByText('tab3'))
  expect(queryByText('this is tab1 content')).toBeInTheDocument()
  expect(queryByText('this is the disable tab-panel')).not.toBeInTheDocument()
})

const TabHooks = () => {
  const tab = useTabs()
  return (
    <ThemeProvider theme={theme}>
      <div>
        <TabList {...tab}>
          <Tab>Tab Hook 1</Tab>
          <Tab>Tab Hook 2</Tab>
        </TabList>
        <TabPanels {...tab}>
          <TabPanel>1 this is the panel of tab hook 1</TabPanel>
          <TabPanel>2 this is the panel of tab hook 2</TabPanel>
        </TabPanels>
      </div>
    </ThemeProvider>
  )
}

test('hooks working', () => {
  const { getByText, queryByText } = render(<TabHooks />)

  expect(queryByText('1 this is the panel of tab hook 1')).toBeInTheDocument()
  expect(
    queryByText('2 this is the panel of tab hook 2')
  ).not.toBeInTheDocument()
  fireEvent.click(getByText('Tab Hook 2'))
  expect(
    queryByText('1 this is the panel of tab hook 1')
  ).not.toBeInTheDocument()
  expect(queryByText('2 this is the panel of tab hook 2')).toBeInTheDocument()
})
