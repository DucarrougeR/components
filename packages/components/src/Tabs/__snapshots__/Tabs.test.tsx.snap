// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs snapshot works as expected 1`] = `ShallowWrapper {}`;
